File: tests/test-minecraft/src/main/java/pize/tests/minecraft/utils/log/Logger.java
package pize.tests.minecraft.utils.log;

import java.io.PrintStream;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

public class Logger{

    private final PrintStream out,err;
    private final DateTimeFormatter dateTimeFormatter;

    private Logger(PrintStream out,PrintStream err){
        this.out = out;
        this.err = err;
        dateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss");
    }

    private Logger(PrintStream out){
        this(out,out);
    }

    private Logger(){
        this(System.out,System.err);
    }


    public void trace(String trace){
        out.println( getLogPrefix(LogLevel.TRACE) + trace );
    }

    public void debug(String debug){
        out.println( getLogPrefix(LogLevel.DEBUG) + debug );
    }

    public void info(String info){
        out.println( getLogPrefix(LogLevel.INFO) + info );
    }

    public void warn(String warn){
        out.println( getLogPrefix(LogLevel.WARN) + warn );
    }

    public void error(String error){
        err.println( getLogPrefix(LogLevel.ERROR) + error );
    }


    public void trace(String trace,Object object){
        out.println( getLogPrefix(LogLevel.TRACE) + trace.replaceAll("\\{}",object.toString()) );
    }

    public void debug(String debug,Object object){
        out.println( getLogPrefix(LogLevel.DEBUG) + debug.replaceAll("\\{}",object.toString()) );
    }

    public void info(String info,Object object){
        out.println( getLogPrefix(LogLevel.INFO) + info.replaceAll("\\{}",object.toString()) );
    }

    public void warn(String warn,Object object){
        out.println( getLogPrefix(LogLevel.WARN) + warn.replaceAll("\\{}",object.toString()) );
    }

    public void error(String error,Object object){
        err.println( getLogPrefix(LogLevel.ERROR) + error.replaceAll("\\{}",object.toString()) );
    }


    private String getLogPrefix(LogLevel level){
        return "[" + ZonedDateTime.now().format(dateTimeFormatter) + "] [" + Thread.currentThread().getName() + "/" + level.name() + "]: ";
    }


    private static final Logger instance = new Logger();

    public static Logger instance(){
        return instance;
    }

}

File: tests/test-minecraft/src/main/kotlin/pize/tests/minecraft/utils/log/Logger.kt
package pize.tests.minecraft.utils.log

import java.io.PrintStream
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

class Logger private constructor(private val out: PrintStream = System.out, private val err: PrintStream = System.err) {
    private val dateTimeFormatter: DateTimeFormatter

    private constructor(out: PrintStream) : this(out, out)

    fun trace(trace: String) {
        out.println(getLogPrefix(LogLevel.TRACE) + trace)
    }

    fun debug(debug: String) {
        out.println(getLogPrefix(LogLevel.DEBUG) + debug)
    }

    fun info(info: String) {
        out.println(getLogPrefix(LogLevel.INFO) + info)
    }

    fun warn(warn: String) {
        out.println(getLogPrefix(LogLevel.WARN) + warn)
    }

    fun error(error: String) {
        err.println(getLogPrefix(LogLevel.ERROR) + error)
    }

    fun trace(trace: String, `object`: Any) {
        out.println(getLogPrefix(LogLevel.TRACE) + trace.replace("\\{}".toRegex(), `object`.toString()))
    }

    fun debug(debug: String, `object`: Any) {
        out.println(getLogPrefix(LogLevel.DEBUG) + debug.replace("\\{}".toRegex(), `object`.toString()))
    }

    fun info(info: String, `object`: Any) {
        out.println(getLogPrefix(LogLevel.INFO) + info.replace("\\{}".toRegex(), `object`.toString()))
    }

    fun warn(warn: String, `object`: Any) {
        out.println(getLogPrefix(LogLevel.WARN) + warn.replace("\\{}".toRegex(), `object`.toString()))
    }

    fun error(error: String, `object`: Any) {
        err.println(getLogPrefix(LogLevel.ERROR) + error.replace("\\{}".toRegex(), `object`.toString()))
    }

    private fun getLogPrefix(level: LogLevel): String {
        return "[" + ZonedDateTime.now()
            .format(dateTimeFormatter) + "] [" + Thread.currentThread().name + "/" + level.name + "]: "
    }

    init {
        dateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss")
    }

    companion object {
        private val instance = Logger()
        fun instance(): Logger {
            return instance
        }
    }
}
